File: 2473-max-sum-of-a-pair-with-equal-sum-of-digits/max-sum-of-a-pair-with-equal-sum-of-digits.py
class Solution:
    def maximumSum(self, nums: List[int]) -> int:
        def helper(i):
            cr = 0 
            while i:
                cr += (i % 10)
                i //= 10 
            return cr 
        
        mx = [0] * 100
        ans = -1 
        for i in nums:
            sm = helper(i)
            if mx[sm] != 0: ans = max(ans, mx[sm] + i)
            mx[sm] = max(mx[sm], i)
        return ans 

File: 2473-max-sum-of-a-pair-with-equal-sum-of-digits/max-sum-of-a-pair-with-equal-sum-of-digits.java
import java.util.*;

class Solution {
    public int maximumSum(int[] nums) {
        // Map to store numbers grouped by their digit sums
        Map<Integer, List<Integer>> mp = new HashMap<>();

        // Group numbers by their digit sums
        for (int num : nums) {
            int sum = sumOfDigits(num);
            mp.computeIfAbsent(sum, k -> new ArrayList<>()).add(num);
        }

        int ans = -1;

        // Iterate through each group in the map
        for (List<Integer> group : mp.values()) {
            // Skip groups with fewer than 2 numbers
            if (group.size() < 2) continue;

            // Sort the group in descending order
            group.sort(Collections.reverseOrder());

            // Calculate the sum of the two largest numbers in the group
            int sum = group.get(0) + group.get(1);
            ans = Math.max(ans, sum);
        }

        return ans;
    }

    // Helper method to calculate the sum of digits of a number
    private int sumOfDigits(int num) {
        int sum = 0;
        while (num > 0) {
            sum += num % 10;
            num /= 10;
        }
        return sum;
    }
}